### PR TITLE
Fixes #37081 - Add a setting to configure PAGE_SIZE

### PR DESCRIPTION
**Pull request by wbclark**

### CODE REVIEW

File: templates/settings.py.erb [hunk 1]
```diff
@@ -52,6 +52,9 @@ WORKING_DIRECTORY = "<%= scope['pulpcore::cache_dir'] %>"
 REMOTE_USER_ENVIRON_NAME = '<%= scope['pulpcore::remote_user_environ_name'] %>'
 AUTHENTICATION_BACKENDS = ['pulpcore.app.authentication.PulpNoCreateRemoteUserBackend']
 
+<% unless scope['pulpcore::page_size'].nil? -%>
+REST_FRAMEWORK__PAGE_SIZE = '<%= scope['pulpcore::page_size'] %>'
```
Comment (evgeni) [inline templates/settings.py.erb:56]: You test for an integer, but here it's quoted like a string, so it fails ;)

**Review by evgeni — APPROVED** [review]
ACK pending template and tests agree ;)

File: templates/settings.py.erb (hunk 1 above)
Comment (wbclark) [reply templates/settings.py.erb]: Fixed!

File: spec/acceptance/settings_spec.rb [hunk 2]
```diff
@@ -51,6 +51,40 @@ class { 'pulpcore':
   end
 end
 
+describe 'PAGE_SIZE setting' do
+  context 'default PAGE_SIZE' do
```
Comment (ekohl) [inline spec/acceptance/settings_spec.rb:55]: I'd keep this as a unit test, just to keep it fast. Acceptance tests are much slower and what are we really testing here? We already have tests for the default case. Testing with a custom page size can make sense, just to make sure the application doesn't crash on start up.

**Review by ekohl — CHANGES_REQUESTED** [review]
A small note on why this is useful would be nice. Though I also wonder if Pulp itself should increase the page size where needed. https://www.django-rest-framework.org/api-guide/pagination/ describes you can also do this for specific views. It also describes a mechanism similar to what Foreman has that allows you to request bigger pages. Has any effort been taken to use that route? It would give a better out of the experience for all users, not just those who tune the installer.

File: spec/acceptance/settings_spec.rb (hunk 2 above)
Comment (wbclark) [reply spec/acceptance/settings_spec.rb]: Good point! I moved the default case into the unit tests, and left only the custom page size test in acceptance.

**Comment by wbclark** [comment]
> A small note on why this is useful would be nice. Though I also wonder if Pulp itself should increase the page size where needed. https://www.django-rest-framework.org/api-guide/pagination/ describes you can also do this for specific views. It also describes a mechanism similar to what Foreman has that allows you to request bigger pages. Has any effort been taken to use that route? It would give a better out of the experience for all users, not just those who tune the installer.

Most users probably don't need to tune this. In some cases, it was helpful as a workaround for other issues, such as one issue that occurred when using Katello to provide content for provisioning OpenStack. In examples like that, the end user couldn't simply request bigger pages because they are using an install script from another project. Having a setting here allows it to be configured persistently so that they aren't blocked when these types of issues occur.

**Comment by ehelms** [comment]
I am curious why this is what I would call a system configuration change. The ability to change this value due to performance issues feels like it should be at the application level to prevent a service restart.

**Comment by ekohl** [comment]
> Most users probably don't need to tune this. In some cases, it was helpful as a workaround for other issues, such as one issue that occurred when using Katello to provide content for provisioning OpenStack. In examples like that, the end user couldn't simply request bigger pages because they are using an install script from another project. Having a setting here allows it to be configured persistently so that they aren't blocked when these types of issues occur.

This is exactly the kind of information that should be presented in a commit message. Then reviewers don't need to guess why it's useful. And anyone reading the source can also retrieve it with `git blame`.

But still, even if the script today may be limited, it can be worthwhile to still start the conversation. Perhaps it can be improved in the future. And as you may have seen, you can tune individual views so Pulp can still increase the default page size for some views if that's beneficial. If the script has implemented paging correctly, they will automatically benefit from larger pages without being corrected. And if it doesn't page correctly, this change would also break it.

**Comment by wbclark** [comment]
> I am curious why this is what I would call a system configuration change. The ability to change this value due to performance issues feels like it should be at the application level to prevent a service restart.

It turns out that Django settings are designed to be immutable.

https://docs.djangoproject.com/en/5.0/topics/settings/#altering-settings-at-runtime

**Comment by wbclark** [comment]
Sorry, didn't mean to close this -- wrong keyboard shortcut.

**Comment by wbclark** [comment]
> But still, even if the script today may be limited, it can be worthwhile to still start the conversation. Perhaps it can be improved in the future. And as you may have seen, you can tune individual views so Pulp can still increase the default page size for some views if that's beneficial. If the script has implemented paging correctly, they will automatically benefit from larger pages without being corrected. And if it doesn't page correctly, this change would also break it.

Indeed, the OpenStack bug in this case was https://bugs.launchpad.net/tripleo/+bug/2028260

**Review by ehelms — APPROVED** [review]

**Comment by wbclark** [comment]
@ekohl if you have a moment, mind taking a fresh look?

File: templates/settings.py.erb [hunk 3]
```diff
@@ -52,6 +52,9 @@ WORKING_DIRECTORY = "<%= scope['pulpcore::cache_dir'] %>"
 REMOTE_USER_ENVIRON_NAME = '<%= scope['pulpcore::remote_user_environ_name'] %>'
 AUTHENTICATION_BACKENDS = ['pulpcore.app.authentication.PulpNoCreateRemoteUserBackend']
 
+<% unless scope['pulpcore::page_size'].nil? -%>
```
Comment (ekohl) [inline templates/settings.py.erb:55]: I prefer keeping conditions simple
```suggestion
<% if scope['pulpcore::page_size'] -%>
```

**Review by ekohl — COMMENTED** [review]
> Indeed, the OpenStack bug in this case was https://bugs.launchpad.net/tripleo/+bug/2028260

If it doesn't do any pagination then adding a page size is never going to work reliably. When you increase the page size to 200 and there are 201 tags, you will miss one.

Since it isn't performance and this isn't solving it I'm leaning to rejecting this PR.

**Comment by ianballou** [comment]
It's certainly hacky that a workaround for the OpenStack bug is to bump up the Pulpcore page size. However, it's slow for OpenStack to fix their issue, but quick and easy for us to release this configurable option.

OpenStack bug aside, I think other folks could benefit from this setting since it is a Pulp configurable. If we even take this Puppet module out of the scope of Katello, wouldn't it be helpful to have more configurable options?  More configurable options would mean more usefulness for people who just want to set up Pulp via Puppet.

**Comment by ekohl** [comment]
Pulp doesn't document the setting (https://docs.pulpproject.org/pulpcore/configuration/settings.html) so it's unclear if they want to support this at all. Just because Django Rest Framework can do it, doesn't mean the Pulp project supports it.

Increasing page sizes can have certain consequences we can't oversee. For example, certain pages may become slow and lead to time outs.

**Comment by ehelms** [comment]
@dralley Can you weigh in here to help us out?

**Comment by dralley** [comment]
I don't consider this setting part of Pulp, it's just there because we use Django REST Framework, and all the settings get merged together. 

We aren't very likely to switch away from Django REST Framework soon so it's stable enough to use as a workaround if you want but I would not rely on this long-term.

**Comment by ekohl** [comment]
> I don't consider this setting part of Pulp

Based on this and my other concerns I'm closing this.

**Pull request closed without merging**